javascript/ql/src/semmle/javascript/StandardLibrary.qll

```diff
@@ -146,3 +146,37 @@ class StringReplaceCall extends DataFlow::MethodCallNode {
     )
   }
 }
+
+/**
+ * A call to `String.prototype.split`.
+ *
+ * We heuristically include any call to a method called `split`, provided it either
+ * has exactly one arguments, or local data flow suggests that the receiver may be a string.
+ */
+class StringSplitCall extends DataFlow::MethodCallNode {
+  StringSplitCall() {
+    this.getMethodName() = "split" and
+    (getNumArgument() = 1 or getReceiver().mayHaveStringValue(_))
+  }
+
+  /**
+   * Gets a string that determines where the string is split.
+   */
+  string getSplitAt() {
+    getArgument(0).mayHaveStringValue(result)
+    or
+    result =
+      getArgument(0).getALocalSource().(DataFlow::RegExpCreationNode).getRoot().getAMatchedString()
+  }
+
+  /**
+   * Gets a the SourceNode for the string before it is split.
+   */
+  DataFlow::SourceNode getUnsplit() { result = getReceiver().getALocalSource() }
```

Comment:
I dislike the `getUnsplit` name (even though it is my own suggestion). 

Perhaps `getStringSource`, and dually `getASubstringRead(int i)` below would be better?

This has no result for `"foo|bar".split("|").map(...)`. I can live with that if a renaming like the above is implemented.
If we need to reason about the shorthand above, we can introduce `string getString()` and `string getSubstring(int i)`.

Reply:
In the `StringOps` module I tried to stick with `getBaseString` as it was meaningful yet generic enough to be used consistently across the different string operations. I think we can use the same here, and possibly drop the `getALocalSource`.

Reply:
I went with `getBaseString`, and dropped the `getALocalSource()`.